src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs

```diff
@@ -97,5 +97,15 @@ internal override ImmutableArray<Diagnostic> GetConstantValueDiagnostics(BoundEx
         {
             return _originalVariable.GetConstantValueDiagnostics(boundInitValue);
         }
+
+        internal TypeSubstitutedLocalSymbol WithSynthesizedLocalKindAndSyntax(SynthesizedLocalKind kind, SyntaxNode syntax)
```

Comment:
Did you consider making `WithSynthesizedLocalKindAndSyntax` an abstract method on LocalSymbol?  That would simplify the rewriter code.